path_role.go

```diff
@@ -0,0 +1,346 @@
+					Type:        framework.TypeString,
+					Description: `The 'sub' claim that is valid for login. Optional.`,
+				},
+				"bound_audiences": &framework.FieldSchema{
+					Type:        framework.TypeCommaStringSlice,
+					Description: `Comma-separated list of 'aud' claims that are valid for login`,
+				},
+				"user_claim": &framework.FieldSchema{
+					Type:        framework.TypeString,
+					Description: `The claim to use for the Identity entity alias name`,
+				},
+				"groups_claim": &framework.FieldSchema{
+					Type:        framework.TypeString,
+					Description: `The claim to use for the Identity group alias names`,
+				},
+				"bound_cidrs": &framework.FieldSchema{
+					Type: framework.TypeCommaStringSlice,
+					Description: `Comma-separated list of IP CIDRS that are allowed to 
+authenticate against this role`,
+				},
+			},
+			ExistenceCheck: b.pathRoleExistenceCheck,
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.CreateOperation: b.pathRoleCreateUpdate,
+				logical.UpdateOperation: b.pathRoleCreateUpdate,
+				logical.ReadOperation:   b.pathRoleRead,
+				logical.DeleteOperation: b.pathRoleDelete,
+			},
+			HelpSynopsis:    strings.TrimSpace(roleHelp["role"][0]),
+			HelpDescription: strings.TrimSpace(roleHelp["role"][1]),
+		},
+	}
+}
+
+type jwtRole struct {
+	// Policies that are to be required by the token to access this role
+	Policies []string `json:"policies"`
+
+	// TokenNumUses defines the number of allowed uses of the token issued
+	NumUses int `json:"num_uses"`
+
+	// Duration before which an issued token must be renewed
+	TTL time.Duration `json:"ttl"`
+
+	// Duration after which an issued token should not be allowed to be renewed
+	MaxTTL time.Duration `json:"max_ttl"`
+
+	// Period, if set, indicates that the token generated using this role
+	// should never expire. The token should be renewed within the duration
+	// specified by this value. The renewal duration will be fixed if the
+	// value is not modified on the role. If the `Period` in the role is modified,
+	// a token will pick up the new value during its next renewal.
+	Period time.Duration `json:"period"`
+
+	// Role binding properties
+	BoundAudiences []string                      `json:"bound_audiences"`
+	BoundSubject   string                        `json:"bound_subject"`
+	BoundCIDRs     []*sockaddr.SockAddrMarshaler `json:"bound_cidrs"`
+	UserClaim      string                        `json:"user_claim"`
+	GroupsClaim    string                        `json:"groups_claim"`
+}
+
+// role takes a storage backend and the name and returns the role's storage
+// entryÍ
+func (b *jwtAuthBackend) role(ctx context.Context, s logical.Storage, name string) (*jwtRole, error) {
+	raw, err := s.Get(ctx, "role/"+strings.ToLower(name))
+	if err != nil {
+		return nil, err
+	}
+	if raw == nil {
+		return nil, nil
+	}
+
+	role := new(jwtRole)
+	if err := json.Unmarshal(raw.Value, role); err != nil {
```

Comment:
`raw.DecodeJSON(role)` is a little tidier, and uses jsonutil.

Reply:
Done